### PR TITLE
fix(core): :ambulance: Fix for calculating durability of a provider

### PR DESCRIPTION
**Pull request by vizio360**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #11002

## What is the new behavior?
Providers are correctly scoped.

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

## Other information

nestJSBug11002 drawio

### CODE REVIEW

File: packages/core/injector/instance-wrapper.ts [hunk 1]
```diff
@@ -203,15 +203,18 @@ export class InstanceWrapper<T = any> {
     if (!isUndefined(this.isTreeDurable)) {
       return this.isTreeDurable;
     }
-    if (this.durable === true) {
-      this.isTreeDurable = true;
-      this.printIntrospectedAsDurable();
+    if (this.scope === Scope.REQUEST) {
+      this.isTreeDurable = this.durable === undefined ? false : this.durable;
```
Comment (SocketSomeone) [inline packages/core/injector/instance-wrapper.ts:207]: Maybe `!!this.durable` or `Boolean(this.durable)` ?
So that the null value is not true?

Comment (SocketSomeone) [reply packages/core/injector/instance-wrapper.ts:207]: oh, if(this.isTreeDurable). my miss

**Comment by vizio360** [comment]
getting a better understanding of the issue, but this PR is not completely fixing the problem. 
Will update once I get my head round it.

**Comment by vizio360** [comment]
Ok, the problems were multiple. 

The issue I found in our code base, and I bet the issue other people had, was caused by the logic in 
`introspectDepsAttribute` combined with the `callback` logic passed in.

For example, to determine if the dependency tree is NON durable, we just need to figure out if at least a dependency, property or enhancer is NON static AND NON durable. This was not working correctly as the check for dependencies, properties and enhancers are done in sequence, and if your enhancers computed as durable, then that would be used as your answer, as the enhancers are the last ones to get checked in `introspectDepsAttribute`.
This caused our issue, as we had a Pipe as a parameter on a function (considered an enhancer) in a controller, and that determined the durability of the entire controller.

Because of the inverse logic I also had to modify the `isDependencyTreeStatic` function, inverting the logic passed into the `introspectDepsAttribute`. Before it was checking that all dependencies had to be static to consider the tree static, but now, similar to durability, I just check if at least one dependency, property or enhancer is non static. If that is the case then the tree is non static.

On top of that I managed to create some tests to cover the scenario I was facing in our code base and some others as well.

**Comment by vizio360** [comment]
@kamilmysliwiec @micalevisk wouldn't mind your input on this one. I did have to invert the logic of the `isDependencyTreeStatic` so now to determine if a dependency, property or enhancer is static I check if at least one of them is non static. If that is the case the tree is NON static, otherwise static.

```
 this.isTreeStatic = !this.introspectDepsAttribute(
      (collection, registry) =>
        collection.some(
          (item: InstanceWrapper) => !item.isDependencyTreeStatic(registry),
        ),
      lookupRegistry,
    );
```

**Comment by vizio360** [comment]
oh and please do take your time to check the tests. Hopefully I covered most of the scenarios now.

**Comment by kamilmysliwiec** [comment]
This looks great, thanks for investigating this issue @vizio360 🙏 🙏

**Comment by kamilmysliwiec** [comment]
> I wonder, do you guys have a beta program? So that new stuff is release under a beta (unstable) version before it gets merge into master? Would provide a way for people to test out a new version before committing to it.

We do release unstable versions under the "next" tag (in the NPM registry) but this one (change) was quite small & wasn't meant to break anything so we went ahead and published it under the latest tag (which might have been a mistake 🥲 )

**Comment by vizio360** [comment]
> > I wonder, do you guys have a beta program? So that new stuff is release under a beta (unstable) version before it gets merge into master? Would provide a way for people to test out a new version before committing to it.
> 
> We do release unstable versions under the "next" tag (in the NPM registry) but this one (change) was quite small & wasn't meant to break anything so we went ahead and published it under the latest tag (which might have been a mistake 🥲 )

yeah sorry about that, I also thought it was a simple fix...

**Comment by davidzwa** [comment]
> This looks great, thanks for investigating this issue @vizio360 🙏 🙏

Yeah I second this, tracked your progress and it was inspiring to see. Well done and thanks a lot!

**Pull request merged**